### PR TITLE
Migrate to circleci 2.0

### PR DESCRIPTION
**Pull request by bliof**
Builds:
  - build - downloads the cc-test-reporter
  - rubocop
  - ruby-2.2
  - ruby-2.3
  - ruby-2.4
  - ruby-2.5
  - upload-coverage
  - push-to-rubygems - only for tags

TODO:
  - [x] gem cache
  - [x] codeclimate + simplecov update
  - [x] fix push-to-rubygems

### CODE REVIEW

**Comment by timstott** [comment]
Excellent work @bliof! Thanks for doing this. I've allowed myself to make a small contribution with Circle CI caching, see #105 🙇

**Comment by bliof** [comment]
@timstott @sgerrand 👀 

We have to change the requirements.

The `codeclimate/diff-coverage` and `codeclimate/total-coverage` most probably rely on having data in master @mattsoutherden had the same issue in another repo and when the PR got merged it resolved itself. I am for removing them as requirements.

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+          command: |
+            mkdir -p tmp/
+            curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64 > ./tmp/cc-test-reporter
+            chmod +x ./tmp/cc-test-reporter
+      - persist_to_workspace:
+          root: ./tmp
+          paths:
+            - cc-test-reporter
+  upload-coverage:
+    working_directory: ~/loga
```
Comment (sgerrand) [inline .circleci/config.yml]: 🙋 Suggest setting this default earlier in the file also. You can `<<:` it into `&defaults` above.

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+jobs:
+  build:
+    working_directory: ~/loga
+    docker:
+      - image: circleci/ruby:2.5.1
+    steps:
+      - run:
+          name: Download cc-test-reporter
+          command: |
+            mkdir -p tmp/
```
Comment (sgerrand) [inline .circleci/config.yml]: 🙋 If the sole purpose of creating `tmp/` is to store the `cc-test-reporter` binary then it's not necessary. You can cache the file using `root: .`.

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+      - image: circleci/ruby:2.5.1
+    steps:
+      - checkout
+      - *restore_gems
+      - *bundle_install
+      - *cache_gems
+      - run:
+          name: Build gem
+          command: gem build $CIRCLE_PROJECT_REPONAME.gemspec
+      - run:
```
Comment (sgerrand) [inline .circleci/config.yml]: 🙋 This step _must_ be a `deploy` key. It's important that it is only attempted once.

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+          name: Publish to rubygems
+          command: |
+            echo ":rubygems_api_key: $RUBYGEMS_API_KEY" >  ~/.gem/credentials
+            chmod 0600 ~/.gem/credentials
+            gem push $CIRCLE_PROJECT_REPONAME-$(echo $CIRCLE_TAG | sed -e 's/v//').gem
+workflows:
+  version: 2
+  test-then-push-gem:
+    jobs:
+      - build
```
Comment (sgerrand) [inline .circleci/config.yml]: 🙅 There's a important thing to remember with CircleCI 2.0 workflows:

> CircleCI treats tag and branch filters differently when deciding whether a job should run.
>
> 1. For a branch push unaffected by any filters, CircleCI runs the job.
> 1. For a tag push unaffected by any filters, CircleCI skips the job.
>
> Item two above means that a job **must** have a `filters tags` section to run as a part of a tag push and all its transitively dependent jobs **must** also have a `filters tags` section.

https://circleci.com/docs/2.0/workflows/#git-tag-job-execution

Comment (sgerrand) [reply .circleci/config.yml]: So in order for the `push-to-rubygems` job to trigger, the `build` job needs to be configured something like:
```yaml
- build:
    filters:
      tags:
        only: /.+/
```

File: Appraisals [hunk 5]
```diff
@@ -1,17 +1,27 @@
-appraise 'rails32' do
-  gem 'rails', '~> 3.2.0'
+if Gem::Version.new(RUBY_VERSION) < Gem::Version.new('2.4.0')
```
Comment (sgerrand) [inline Appraisals]: 👏

File: .circleci/config.yml (hunk 2 above)
Comment (bliof) [reply .circleci/config.yml]: I would like not to pollute the working directory with some random scripts. There isn't a `checkout` here but we could add one in the future 🤷‍♂️

Also everywhere it is the workspace is attached at `tmp` and it feels less magical to me with this line :)

File: .circleci/config.yml (hunk 3 above)
Comment (bliof) [reply .circleci/config.yml]: What is `deploy` key? I am lost on this one.

Comment (timstott) [reply .circleci/config.yml]: @sgerrand I don't know that the deploy key provides that guarantee. According to the [documentation](https://circleci.com/docs/2.0/configuration-reference/#deploy) its only concerns jobs that run in parallel.

> In general deploy step behaves just like run with one exception - in a job with parallelism, the deploy step will only be executed by node #0 and only if all nodes succeed. Nodes other than #0 will skip this step.

Comment (bliof) [reply .circleci/config.yml]: So it is in the doc.. Those search boxes at the top are pointless 🙈

File: spec/spec_helper.rb [hunk 6]
```diff
@@ -6,7 +6,9 @@
 require 'rack/test'
 require 'simplecov'
 
-SimpleCov.start
+SimpleCov.start do
+  command_name "ruby-#{RUBY_VERSION}-#{File.basename(ENV['BUNDLE_GEMFILE'], '.gemfile')}"
```
Comment (timstott) [inline spec/spec_helper.rb]: Very noice!

**Comment by bliof** [comment]
@sgerrand 👀

**Pull request merged**